### ISSUE TITLE
Use largest frame on page for scrolling in case for multiple pages

### ISSUE THREAD

**Issue by jgkamat** [The-Compiler/qutebrowser#1354]
"frame-flattening" is very nice, but in some situations where there are multiple frames on a page, it would be nicer if we could scroll the largest frame on the page using the scrolling commands rather than not being able to scroll at all.

Frame flattening seems to work well until a class is selected in the java api, at which point scrolling breaks.

See [the java api](http://docs.oracle.com/javase/8/docs/api/) for an example of this in action.

I'm using the latest stable.

**Comment by lahwaacz** [comment]
The largest frame on a page is always the "master" frame, there is no point in scrolling it since if there are some sub-frames, the "master" frame usually fits exactly to the screen. The frame selected for scrolling depends on focus, which is exactly the same behaviour as other browsers use for scrolling shortcuts (in fact, I think that QtWebKit is responsible for it). If there is a need for better heuristics, I think that automatically moving focus to the frame where a link has been opened might help.

**Comment by jgkamat** [comment]
Hmm I didn't notice that the focus seemed to work before. It does work for me if I use `:scroll down`, but If I try using `:scroll-page 0 -1` it dosen't seem to do anything. Perhaps this is an issue only in `scroll-down`? I switched because I can't find a way to customize scrolling amounts with `:scroll down`, but with `:scroll-page` I can do so.

**Comment by The-Compiler** [comment]
This is unfortunately a limitation of QtWebKit's API - `:scroll down` simply emulates a cursor down keypress to circumvent that.

I'll take another look at scrolling issues after QtWebEngine support (#666) is ready, there might be a better solution there.

**Comment by amerlyq** [comment]
https://weechat.org/files/doc/stable/weechat_user.en.html
There are left and right frames: when scrolling, the left one is always preffered and scrolled till limit, before scrolling the right one. Therefore, scrolling direction can't be changed immediately, until left frame completely scrolled. Issue seems related here. Point me, if it requires separate issue.

**Comment by The-Compiler** [comment]
@amerlyq The right one is scrolled here. What backend are you using? Does clicking it help? Does `:bind j` show `scroll down`?

**Comment by amerlyq** [comment]
You have the point. Seems like I jumped by one of links on left frame, after which the aforementioned behavior was observed. Clicking right frame by mouse restored normal behavior -- I didn't checked that immediately, having mouse plugged out for eons. Anyway, it's a pity, I couldn't find the way to move focus to right frame w/o the mouse (currently emulated by accessibility xkb on numpad).

**Comment by The-Compiler** [comment]
vimperator [has some code](https://github.com/vimperator/vimperator-labs/blob/master/common/content/buffer.js#L1077) for this:

```js
    findScrollableWindow: function findScrollableWindow() {
        let win = this.focusedWindow;
        if (win && (win.scrollMaxX > 0 || win.scrollMaxY > 0))
            return win;
        win = config.browser.contentWindow;
        if (win.scrollMaxX > 0 || win.scrollMaxY > 0)
            return win;
        for (let frame in util.Array.itervalues(win.frames))
            if (frame.scrollMaxX > 0 || frame.scrollMaxY > 0)
                return frame;
        return win;
    },
    findScrollable: function findScrollable(dir = 0, horizontal) {
        let pos = "scrollTop", maxPos = "scrollTopMax", clientSize = "clientHeight";
        if (horizontal)
            pos = "scrollLeft", maxPos = "scrollLeftMax", clientSize = "clientWidth";
        function find(elem) {
            if (elem && !(elem instanceof Element))
                elem = elem.parentNode;
            for (; elem && elem.parentNode instanceof Element; elem = elem.parentNode) {
                if (elem[clientSize] == 0)
                    continue;
                if (dir < 0 && elem[pos] > 0 || dir > 0 && elem[pos] < elem[maxPos] || dir == 0 && elem[maxPos] > 0)
                    break;
            }
            return elem;
        }
        let win = this.focusedWindow;
        if (win.getSelection().rangeCount)
            var elem = find(win.getSelection().getRangeAt(0).startContainer);
        if (!(elem instanceof Element)) {
            let doc = Buffer.findScrollableWindow().document;
            elem = find(doc.body || doc.getElementsByTagName("body")[0] ||
                        doc.documentElement);
        }
        return elem;
    },
```

**Comment by jcjolley** [comment]
This not working in Jira is preventing me from using this as my daily browser :(

**Comment by The-Compiler** [comment]
No issues with scrolling in JIRA here FWIW.

**Comment by jgkamat** [comment]
This is implemented in [vimium here](https://github.com/philc/vimium/blob/026c90ccff6f40cdc370c11339cdfab4a5e6b7b9/content_scripts/scroller.coffee#L98-L114), I'll look into it after #3691 (and maybe creating a hint to select scrollable frames). However, that method will only work with `:scroll-page/px`.

**Comment by reaysawa** [comment]
This is pretty curious - if you search on StackOverflow, most answers regarding scrolling tell you to detect scrollability with `el.scrollHeight > el.clientHeight` or something of the sort.

In the Javadoc pages this doesn't work because the iframe is the one who's actually gating the space taken in the page (what would be the "scrollable" list element is displayed in it's entirely inside the iframe, so it doesn't actually scroll).

![image](https://user-images.githubusercontent.com/17429390/54129554-4a0a2600-43ed-11e9-9c59-1fa34df58e07.png)

From what I've been changing in the code for scrollability, it seems like the missing piece is to consider the topmost embedded frame element (if any) and compare it to the element's size, then attempt to scroll the nearest scrollable parent element (what's being done at https://github.com/qutebrowser/qutebrowser/pull/3691/files#diff-b3fda98a71945f975eff48705e3cfdfdR127).

@jgkamat Are there any more specific cases like this?

**Comment by jgkamat** [comment]
João Paulo writes:

> From what I've been changing in the code for scrollability, it seems like the missing piece is to consider the topmost embedded frame element (if any) and compare it to the element's size, then attempt to scroll the nearest scrollable parent element (what's being done at https://github.com/qutebrowser/qutebrowser/pull/3691/files#diff-b3fda98a71945f975eff48705e3cfdfdR127).
>
> @jgkamat Are there any more specific cases like this?

I was planning on sticking as close to vimiums implementation as possible.
That procedure is used in vimium appears to be:

1. Check to see if the document's scrollable element is really scrollable
(using the scrolling method described above, and in the PR), if so, use it.
2. Iterate through all the children (recursively), sorted by the size of the
element, and return the first one that scrolls (according to the previous
definition).

I'm assuming the iteration is slow if there's no matches found, but if there's
no matches found, the DOM should be pretty small anyway (and if it works for
vimium, I'm sure it will work for us too).

I linked to vimium's code in the previous comment already.

**Comment by chaorace** [comment]
Since this PR is probably on the backburner, I went ahead and bodged a userscript which does the thing. It's all based on the Vimperator code previously shared in this thread and should be more-or-less functionally identical. I was a bit lazy and only set it up for vertical scrolling, though.

<details>
<summary><b>Usage Summary</b> (toggle)</summary>

A `scrollHelper` JS object is populated in the window scope of the jseval world by the provided Userscript. This means that the object and associated functions are only accessible via Qutebrowser's `:jseval`. You must rebind your keys to invoke these functions in order to take advantage of this fix. An example configuration is included in final section.

There are 4 functions provided:

- `scrollHelper.scrollTo(verticalPixelCount)`: Scrolls to an absolute vertical position (units are CSS pixels)
- `scrollHelper.scrollToPercent(percentageInteger)`: Equivalent to `:scroll-to-perc`. Scrolls to an absolute vertical position (1 unit = 1%)
- `scrollHelper.scrollBy(verticalScrollDistance)`: Equivalent to `:scroll-px`. Scrolls down by the given amount (units are CSS pixels, negative values go up)
- `scrollHelper.scrollPage(verticalScrollPages)`: Equivalent to `:scroll-page`. Scrolls down by the given amount (1 unit = 1 vertical page, negative values go up)

</details>

<details>
<summary><b>Userscript Code</b> (toggle)</summary>

```js
// ==UserScript==
// @name JS-based scrolling that bypasses QB frame selection issues
// @qute-js-world jseval
// @run-at document-start
// ==/UserScript==
unsafeWindow.scrollHelper = (() => {
    const getFocusedWindow = (nextElem) => {
        if (nextElem === null) return null
        if (nextElem === undefined) return getFocusedWindow(window.document.activeElement ?? null)
        return getFocusedWindow(nextElem.contentDocument?.activeElement ?? null) ?? nextElem.ownerDocument?.defaultView ?? null
    }

    const getScrollMaxY = ({ document: { documentElement } }) => documentElement.scrollHeight - documentElement.clientHeight

    const getWindowVisibleArea = ({ document: { documentElement } }) => documentElement.clientHeight * documentElement.clientWidth

    const findVertScrollableWindow = () => {
        const focusedWindow = getFocusedWindow() ?? window
        if (getScrollMaxY(focusedWindow) > 0) return focusedWindow
        if (getScrollMaxY(window) > 0) return window

        return Array
            .from(window.frames)
            .sort((x, y) => getWindowVisibleArea(y) - getWindowVisibleArea(x))
            .find((frame) => getScrollMaxY(frame) > 0) ?? window
    }

    const getScrollTopMax = (elem) => elem.scrollHeight - elem.clientHeight

    const findVertScrollableAncestor = (delta, nextElem) => {
        if (!(nextElem?.parentNode instanceof Element)) return nextElem

        if (nextElem.clientHeight !== 0) {
            if (delta < 0 && nextElem.scrollTop > 0) return nextElem
            if (delta > 0 && nextElem.scrollTop < getScrollTopMax(nextElem)) return nextElem
            if (delta === 0 && getScrollTopMax(nextElem) > 0) return nextElem
        }

        return findVertScrollableAncestor(delta, nextElem.parentNode)
    }

    const getSelectionElem = () => {
        const selection = getFocusedWindow().getSelection()
        return selection.rangeCount !== 0
            ? selection.getRangeAt(0).startContainer
            : null
    }

    const getParentIfNotElement = (maybeElement) => maybeElement instanceof Element ? maybeElement : maybeElement?.parentNode

    const findVertScrollable = (delta = 0) => {
        const selectionScrollableElem = findVertScrollableAncestor(delta, getParentIfNotElement(getSelectionElem()))
        if (selectionScrollableElem instanceof Element) return selectionScrollableElem

        const scrollableDoc = findVertScrollableWindow().document
        const scrollableElem = scrollableDoc.body || scrollableDoc.getElementsByTagName("body")[0] || scrollableDoc.documentElement
        return findVertScrollableAncestor(delta, getParentIfNotElement(scrollableElem))
    }

    return {
        scrollTo: (position) => findVertScrollable().scrollTo({top: position}),
        scrollToPercent: (percentPosition) => {
            const scrollElement = findVertScrollable()
            const paneHeight = scrollElement.scrollHeight
            scrollElement.scrollTo({top: percentPosition / 100 * paneHeight})
        },
        scrollBy: (delta) => findVertScrollable(delta).scrollBy({top: delta, behavior: 'smooth'}),
        scrollPage: (pages) => {
            const fakeDelta = pages < 0 ? -10 : 10
            const scrollElement = findVertScrollable(fakeDelta)
            const pageHeight = scrollElement.clientHeight
            scrollElement.scrollBy({top: pageHeight * pages})
        },
    }

})()
```
</details>

<details>
<summary><b>Example Binds/Configuration</b> (toggle)</summary>

You **must** rebind your scroll keys to the `scrollHelper` functions in order to take advantage of the Userscript, nothing will change otherwise.
```python
config.bind('j', 'jseval --quiet scrollHelper.scrollBy(200)')
config.bind('k', 'jseval --quiet scrollHelper.scrollBy(-200)')
config.bind('<Ctrl-D>', 'jseval --quiet scrollHelper.scrollPage(0.8)')
config.bind('<Ctrl-U>', 'jseval --quiet scrollHelper.scrollPage(-0.8)')
config.bind('gg', 'jseval --quiet scrollHelper.scrollTo(0)')
config.bind('G', 'jseval --quiet scrollHelper.scrollToPercent(100)')
```
</details>

**Comment by chaorace** [comment]
Heads up to anyone who's been using my `scrollHelper` script: I've updated it with the following changes...
* Fixed failure to scroll that could happen if there was no active element AND none of the window frames were scrollable
* Shadow DOM content can now be scrolled (assuming the shadow DOM tree isn't closed)

The fix in the first bullet required adding manual DOM traversal code. This codepath is relatively very slow, so we only use it when we _absolutely_ have to. For safety, the traverser stops going down at 10 levels. This limit can be configured by editing the IIFE argument at the bottom of the script. If desired, it's also possible to disable this codepath by setting the depth limit to 0, effectively restoring the old behavior.

Find the script in my above original comment. Prior versions can always be found by viewing the comment's edit history
